### PR TITLE
add debian repo support

### LINKED ISSUE #13: Add repository for Debian/Ubuntu
Currently, the install class only uses a repository for EL: https://github.com/puppetlabs/influxdb/blob/7a800f36e6ac49b2060bef957e66a4aa65a07c38/manifests/install.pp#L75-L84

Since Influxdata provides [repositories](https://repos.influxdata.com/) for Ubuntu and Debian, we should add support for them.

### PR DESCRIPTION
**Pull request by SimonHoenscheid**
* regarding to documentation Influxdata only uses one GPG key to sign paackages
fixes #13

### CODE REVIEW

**Comment by m0dular** [comment]
Thank you for opening these PRs!  I'll start taking a look at them today.

**Review by m0dular — CHANGES_REQUESTED** [review]
Thanks for this one, I've been meaning to get to this issue.  The code looks good, but like #36 could we only have the commits related to the feature in the PR, d362b1f9 in this case?

**Comment by SimonHoenscheid** [comment]
@m0dular I included the debian repo support feature here, because it contains variables used by this feature.

**Comment by m0dular** [comment]
Ah that's right, they both use the repository related variables.  I'll kick off the acceptance tests, and I'll find some time to add a couple unit tests to cover it.  Unless you feel like doing those yourself :P

File: manifests/init.pp
Comment (m0dular) [inline manifests/init.pp:87]: I think we need to keep the `and` condition here and add `Debian` to the array, so

```
if $manage_repo and $facts['os']['family'] in ['Redhat', 'Debian'] {
```

In the tests it's trying to use a repository for SLES installs, but I don't think they maintain a repo for that platform.  So we need to default to using an archive source for SLES.

**Review by m0dular — CHANGES_REQUESTED** [review]

**Comment by SimonHoenscheid** [comment]
@m0dular The repo descriptions just references one key:

https://portal.influxdata.com/downloads/

File: manifests/init.pp
Comment (SimonHoenscheid) [reply manifests/init.pp:87]: I would like to keep the condition short and add a `data/os/SLES.yaml` with the value `influxdb::manage_repo: false`

File: hiera.yaml
Comment (bastelfreak) [inline hiera.yaml:12]: that's redundant with line 19?

File: data/os/CentOS.yaml
Comment (bastelfreak) [inline data/os/CentOS.yaml:1]: we can yank this file because RedHat.yaml matches for Centos/Rock/Alma systems as well, because of the identical OS family

File: manifests/init.pp
Comment (bastelfreak) [reply manifests/init.pp:87]: I think only validating $manage_repo here is fine, because there's a case statement afterwards for the individual OS families. Personally I would set $manage_repo to false in common.yaml and enable it only for Debian and RedHat OS family.

File: hiera.yaml
Comment (SimonHoenscheid) [reply hiera.yaml:12]: yes.

File: data/os/CentOS.yaml
Comment (SimonHoenscheid) [reply data/os/CentOS.yaml:1]: The URL is different: `https://repos.influxdata.com/rhel` vs. `https://repos.influxdata.com/centos` thats why I go with the name here.

**Comment by bastelfreak** [comment]
I took a look at the gpg keys:
```
bastelfreak@bastelfreak-nb /tmp $ md5sum influxdb*
e7b6670153b90766ae2f5dc7bdc61c6a  influxdb2.key
e38c47ef9138d841ee94f471f0ab01fb  influxdb.key
e38c47ef9138d841ee94f471f0ab01fb  influxdb.key.old
bastelfreak@bastelfreak-nb /tmp $
```

I didn't find an official statement from influxdb which key has which purpose, but:
* https://docs.influxdata.com/influxdb/v2.0/install/?t=Linux are install instructions for the first influxdb2 release, they use influxdb2.key
* https://docs.influxdata.com/influxdb/v1.8/introduction/install/ are the last influxdb 1 install instructions and use influxdb.key
* I guess they created a new key for the new major release and this repo adds both keys to support influxdb1->2 upgrades?
* I think this PR should keep both keys
* If we want to remove the old key, it should happen in another PR

File: manifests/init.pp
Comment (m0dular) [reply manifests/init.pp:87]: So I misread the failure in the spec tests, my bad.  The actual failure is that `$repo_url` is unset on SLES because it isn't in `common.yaml` and it's a required parameter.  https://github.com/puppetlabs/influxdb/runs/7993767980?check_suite_focus=true#step:10:183

We could make this an `Optional[String]` that defaults to `undef`, or put a default value in `common.yaml`.  I think I like `Optional` more because there isn't one for SLES.

I agree that only checking `$manage_repo` is fine since I was mistaken about the issue.

File: data/os/CentOS.yaml
Comment (m0dular) [reply data/os/CentOS.yaml:1]: Right, they do actually have different base URLs for RedHat and CentOS.  I don't know if it would cause any actual failures if we removed this file, but I think if they offer them we might as well use them.

**Comment by m0dular** [comment]
The acceptance tests would show us whether removing the gpg key would be an issue, but I also think we should keep the GPG url the same for now.  That can be revisited later if we need to.

File: manifests/init.pp
Comment (m0dular) [reply manifests/init.pp:87]: Also, if you wanted to move the defaults for `$manage_repo` to module data that would be fine :+1:

**Comment by SimonHoenscheid** [comment]
@m0dular Only thing to think about: Yumrepo doesn't seem to have a Problem with two keys AFAIK I dont know if Apt has. so my suggestion would be to go with the old key for Ubuntu and Debian

**Comment by m0dular** [comment]
Sounds good.  If you can push a commit that fixes the `$repo_url` error and adds the key back I'll kick off the acceptance tests.  Should we have both keys in `common.yaml` and add the old key to `Debian.yaml` and `Ubuntu.yaml`?

**Comment by SimonHoenscheid** [comment]
Sounds good. will do.

**Comment by SimonHoenscheid** [comment]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------

**Comment by SimonHoenscheid** [comment]
@m0dular PR updated :-)

**Comment by MartyEwings** [comment]
I think the failures a related to the fact that the Debian an ubuntu repos for influx db only keep the latest version in them, we may need to specify latest or something for these OS's

**Comment by SimonHoenscheid** [comment]
@MartyEwings I will add that to the data

**Comment by SimonHoenscheid** [comment]
@MartyEwings  @m0dular done. Do you want me to lint the code or should I rebase after #31  merged? I am also fine with closing #31 and linting #36 and #37

File: manifests/init.pp
Comment (m0dular) [inline manifests/init.pp:126]: There were some errors in the acceptance tests related to the ssl files being applied before the `/etc/influxdb` directory exists, and `influxdb.service.d` being applied before the `influxdb` user exists.  I don't know why this happens on Ubuntu and not elsewhere, but I tested this change and it worked.

```
    $package_before = if $use_ssl and $manage_ssl {
      [
        File['/etc/influxdb/cert.pem', '/etc/influxdb/key.pem', '/etc/influxdb/ca.pem', '/etc/systemd/system/influxdb.service.d'],
        Service['influxdb']
      ]
    }
    else {
      Service['influxdb']
    }

    package {'influxdb2':
      ensure  => $version,
      require => $package_require,
      before  => $package_before,
    }
```

**Review by m0dular — CHANGES_REQUESTED** [review]

**Comment by m0dular** [comment]
I went ahead and merged #31 because it's only linting.  That created a merged conflict here on the closing `)` of the parameters list for... some reason.

For #36 we could either apply the same `$repo_url` and `influxdb::repo_gpg_key_url` changes there and merge it, or just have those changes in this PR.  Whichever way is easier works for me.

**Comment by SimonHoenscheid** [comment]
This PR is waiting for #36 to be merged. Afterwards this PR can be rebased.

**Comment by SimonHoenscheid** [comment]
I already rebased this PR. It can be merged after #36

File: manifests/init.pp
Comment (m0dular) [reply manifests/init.pp]: Hi @SimonHoenscheid, we're still seeing these failures in acceptance tests.

```
Error: Could not set 'file' on ensure: No such file or directory - A directory component in /etc/influxdb/cert.pem20220913-19386-1aod2be.lock does not exist or is a dangling symbolic link (file: /etc/puppetlabs/code/environments/production/modules/influxdb/manifests/init.pp, line: 200)
```

Can you try adding this code so the package is installed before we try to set up the certs?

**Comment by m0dular** [comment]
I added a commit that includes the `$package_require` code and an update to the spec tests, merged in #40.  Thanks for the contribution!

**Pull request merged**
